src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.cs

```diff
@@ -384,5 +387,42 @@ public IEnumerable<string> GetParameterNamesForMethod(MethodDesc method)
                 yield return ecmaMethod.MetadataReader.GetString(p.Name);
             }
         }
+
+        public bool IsTypeDefinedExternally(TypeDesc type)
+        {
+            if (!_multifileCompilation)
+                return false;
+
+            if (type.ContainsGenericVariables)
+                return false;
+
+            if (!(type is MetadataType))
+                return false;
+
+            MetadataType metadataType = type as MetadataType;
+
+            if (metadataType.Module is EcmaModule)
+            {
+                EcmaModule module = metadataType.Module as EcmaModule;
+                string fileName;
+                if (!InputFilePaths.TryGetValue(module.GetName().Name, out fileName))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public bool IsMethodDefinedExternally(MethodDesc method)
+        {
+            if (!_multifileCompilation)
+                return false;
+
+            if (method.GetTypicalMethodDefinition().ContainsGenericVariables || method.OwningType.ContainsGenericVariables)
```

Comment:
`method.OwningType.ContainsGenericVariables` check is duplicated from `IsTypeDefinedExternally`. Seems unnecessary.